e2etests/tests/gracefulrestart.go

```diff
@@ -0,0 +1,128 @@
+	var (
+		cs       clientset.Interface
+		updater  *config.Updater
+		reporter *k8sreporter.KubernetesReporter
+		nodes    []corev1.Node
+	)
+
+	ginkgo.BeforeEach(func() {
+		var err error
+		updater, err = config.NewUpdater()
+		Expect(err).NotTo(HaveOccurred())
+		reporter = dump.NewK8sReporter(k8s.FRRK8sNamespace)
+
+		for _, c := range infra.FRRContainers {
+			err := c.UpdateBGPConfigFile(frrconfig.Empty)
+			Expect(err).NotTo(HaveOccurred(), "clear any previous config in the infra container failed")
+		}
+		err = updater.Clean()
+		Expect(err).NotTo(HaveOccurred(), "clear any previous config in the API failed")
+
+		cs = k8sclient.New()
+		nodes, err = k8s.Nodes(cs)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	ginkgo.AfterEach(func() {
+		if ginkgo.CurrentSpecReport().Failed() {
+			testName := ginkgo.CurrentSpecReport().LeafNodeText
+			dump.K8sInfo(testName, reporter)
+			dump.BGPInfo(testName, infra.FRRContainers, cs)
+		}
+	})
+
+	ginkgo.Context("When restarting the frrk8s deamon pods", func() {
+
+		assertFuncPerEntry := func(ipFam ipfamily.Family, prefix string) {
+			frrs := config.ContainersForVRF(infra.FRRContainers, "")
+			for _, c := range frrs {
+				err := container.PairWithNodes(cs, c, ipFam)
+				Expect(err).NotTo(HaveOccurred(), "set frr config in infra containers failed")
+			}
+
+			peersConfig := config.PeersForContainersWithOptions(frrs, ipFam,
+				config.AllowAllEnabled, config.GracefulRestartEnabled)
+
+			frrConfigCR := frrk8sv1beta1.FRRConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "graceful-restart-test",
+					Namespace: k8s.FRRK8sNamespace,
+				},
+				Spec: frrk8sv1beta1.FRRConfigurationSpec{
+					BGP: frrk8sv1beta1.BGPConfig{
+						Routers: []frrk8sv1beta1.Router{
+							{
+								ASN:       infra.FRRK8sASN,
+								Neighbors: peersConfig.FRRK8sNeibghbor(),
+								Prefixes:  []string{prefix},
+							},
+						},
+					},
+				},
+			}
+
+			err := updater.Update(peersConfig.Secrets, frrConfigCR)
+			Expect(err).NotTo(HaveOccurred(), "apply the CR in k8s api failed")
+
+			check := func() error {
+				for _, p := range peersConfig.Peers() {
+					found, err := routes.CheckNeighborHasPrefix(p.FRR, prefix, nodes)
+					Expect(err).NotTo(HaveOccurred())
+					if !found {
+						return fmt.Errorf("Neigh %s does not have prefix %s", p.FRR.Name, prefix)
+					}
+				}
+				return nil
+			}
+
+			Eventually(check, time.Minute, time.Second).ShouldNot(HaveOccurred(),
+				"route should exist before we restart speaker")
+
+			go func() { // go restart speakers while Consistently check that route exist
+				defer ginkgo.GinkgoRecover()
+				err := cs.CoreV1().Pods(k8s.FRRK8sNamespace).DeleteCollection(context.Background(), metav1.DeleteOptions{}, metav1.ListOptions{
+					LabelSelector: k8s.FRRK8sLabelSelector,
+				})
+				Expect(err).NotTo(HaveOccurred())
+				//DISCUSS: should we trust the delete happens or should we have logic to verify it?
```

Comment:
Not only we must ensure the pods got deleted, we also need to check that the new instances are up before we move up to the next test.

Reply:
Can we elaborate more:
what scenario the delete returns without error but pods are not actually deleted?

why we "need to check new instances are up" is that because other test will fail or you think is important for this test?